src/Workspaces/Core/Portable/Shared/Extensions/SourceTextExtensions.cs

```diff
@@ -229,20 +233,103 @@ private static void WriteChunksTo(SourceText sourceText, ObjectWriter writer, in
-    private sealed class ObjectReaderTextReader : TextReaderWithLength
+    public static SourceText CreateSourceText(
+        SyntaxNode node, Encoding? encoding, SourceHashAlgorithm checksumAlgorithm, CancellationToken cancellationToken)
+    {
+        // If this node is small enough to not go into the LOH, we can just fast path directly to creating a SourceText from it.
+        var totalLength = node.FullWidth();
+        if (totalLength <= SourceTextLengthThreshold)
+            return SourceText.From(node.ToFullString(), encoding, checksumAlgorithm);
+
+        // Allocate the space to write the node into.  Explicitly chunked so that nothing goes into the LOH.
+        var chunks = CreateChunks(totalLength);
+
+        // Write the node into that temp space.
+        using var chunkWriter = new CharArrayChunkTextWriter(totalLength, chunks, encoding!, cancellationToken);
+        node.WriteTo(chunkWriter);
+        Contract.ThrowIfTrue(totalLength != chunkWriter.Position);
+
+        // Call into the text service to make us a SourceText from the chunks.  Disposal of this reader will free all
+        // the intermediary chunks we allocated.
+
+        Contract.ThrowIfTrue(chunks.Any(static (c, s) => c.Length != s, CharArrayLength));
+
+        using var chunkReader = new CharArrayChunkTextReader(chunks, totalLength);
+        var result = SourceText.From(chunkReader, totalLength, encoding, checksumAlgorithm);
+        Contract.ThrowIfTrue(totalLength != chunkReader.Position);
+
+        return result;
+
+        static ImmutableArray<char[]> CreateChunks(int totalLength)
+        {
+            var numberOfChunks = 1 + (totalLength / CharArrayLength);
+            var buffer = new FixedSizeArrayBuilder<char[]>(numberOfChunks);
+            for (var i = 0; i < numberOfChunks; i++)
+                buffer.Add(s_charArrayPool.Allocate());
+
+            return buffer.MoveToImmutable();
+        }
+    }
+
+    private static int GetIndexFromPosition(int position) => position / CharArrayLength;
+    private static int GetColumnFromPosition(int position) => position % CharArrayLength;
+
+    private sealed class CharArrayChunkTextWriter(
+        int totalLength, ImmutableArray<char[]> chunks, Encoding encoding, CancellationToken cancellationToken) : TextWriter
+    {
+        private readonly int _totalLength = totalLength;
+        private readonly ImmutableArray<char[]> _chunks = chunks;
+        private readonly CancellationToken _cancellationToken = cancellationToken;
+
+        /// <summary>
+        /// Public so that caller can assert that writing out the text actually wrote out the full text of the node.
+        /// </summary>
+        public int Position { get; private set; }
+
+        public override Encoding Encoding { get; } = encoding;
+
+        public override void Write(string? value)
+        {
+            Contract.ThrowIfNull(value);
+
+            var valueSpan = value.AsSpan();
+            while (valueSpan.Length > 0)
```

Comment:
>while (valueSpan.Length > 0) [](http://example.com/codeflow?start=12&length=28)

feel free to ignore this, but I find something like the following a bit easier to read:

```
var index = 0;
foreach (var chunk in _chunks)
{
    var copyLength = Math.Min(value.Length - index, chunk.Length);
    value.CopyTo(index, chunk, 0, copyLength);

    index += chunk.Length;
}
```

Reply:
That seems not correct since you're restarting at the first chunk each time a string is written in.

Reply:
oh, duh, multiple write calls

Reply:
ok, my current rewrite of this doesn't really look simpler than what you've got